v21.2/performance-best-practices-overview.md

```diff
@@ -323,53 +323,69 @@ If you have long-running queries (such as analytics queries that perform full ta
 
-## Understanding and avoiding transaction contention
+## Hot spots
 
-Transaction contention occurs when the following three conditions are met:
+Transactions that operate on the same range but _different index key values_ will be limited by the overall hardware capacity of a single node (the range lease holder). These are referred to as _hot spots_.
 
-- There are multiple concurrent transactions or statements (sent by multiple clients connected simultaneously to a single CockroachDB cluster).
-- They operate on the same data, specifically over table rows with the same index key values (either on [primary keys](primary-key.html) or secondary [indexes](indexes.html)) or using index key values that are close to each other, and thus place the indexed data on the same [data ranges](architecture/overview.html).
-- At least one of the transactions modify the data.
+Hot spots can occur when a range is indexed on a column of data that is sequential in nature such that all incoming writes to the range will be the last (or first) item in the index and appended to the end of the range. As a result, the system cannot find a point in the range that evenly divides the traffic, and the range cannot benefit from load-based splitting, creating a hotspot on the single range.
 
-A set of transactions that all contend on the same keys will be limited in performance to the maximum processing speed of a single node (limited horizontal scalability). Non-contended transactions are not affected in this way.
+Read hot spots can occur if you perform lots of scans of an empty table or of a single key.
+
+### Fix hot spots
+
+To reduce hot spots:
+
+- Use index key values with a random distribution of values, so that transactions over different rows are more likely to operate on separate data ranges. See the [SQL FAQs](sql-faqs.html) on row IDs for suggestions.
+
+- Increase [normalization](https://en.wikipedia.org/wiki/Database_normalization) of the data to place parts of the same records that are modified by different transactions in different tables. Fully normalized data allows separate transactions to modify related underlying data without causing contention. Increasing normalization has drawbacks as well as benefits, however. Denormalizing data can increase performance for certain queries by creating multiple copies of often-referenced data in separate ranges, so read-heavy workloads may have higher performance. However, denormalization adds complexity the data model, increases the chance of data inconsistency, increases data redundancy, and can have poor performance when running write-heavy workloads.
+
+- If the application strictly requires operating on very few different index key values, consider using [`ALTER ... SPLIT AT`](split-at.html) so that each index key value can be served by a separate group of nodes in the cluster.
+
+- To avoid read hot spots, use global tables and increase the number of replicas.
```

Comment:
I'm not sure the advice in this bullet is correct.

Using [`GLOBAL` tables](https://www.cockroachlabs.com/docs/v21.1/multiregion-overview.html#global-tables) and [increasing the number of replicas](https://www.cockroachlabs.com/docs/stable/configure-replication-zones.html#num_replicas) are different things.

Re: global tables:

- one can only use global tables in a [multi-region cluster](https://www.cockroachlabs.com/docs/v21.1/multiregion-overview.html)
- global tables are only good for read-mostly workloads on data that must be accessible 
- ... and one should generally not adjust the number of replicas on a `GLOBAL` table, the whole point of the multi-region abstractions is that users should not need to do low-level replication changes
- in general I don't think we should recommend that the median user with a read hot spot should turn on multiregion + global tables without knowing more about their workload / deployment

Re: increasing replication factor:

if you are not running a "new style" multi-region cluster (link above) then I think increasing the number of replicas could make sense, but it's not straightforward and involves tradeoffs.  I'm not the expert on those tradeoffs, but AIUI: more replication costs (CPU for replication queue etc), more storage (for ranges on different nodes), etc.

my understanding is that it's not something we should recommend to users without caveats.  Not sure if we should suggest it to the median user since it could result in outcomes on their cluster they are not prepared for - I would describe it as more intermediate/advanced usage that most users should not do unless they really know what they're doing

Re: read hot spots in general

- another option for avoiding read hot spots in a multi-region deployment is to use [follower reads](https://www.cockroachlabs.com/docs/v21.1/follower-reads.html)
- we also have a [hot ranges API endpoint](https://www.cockroachlabs.com/docs/v21.2/cluster-api#resources) which is used to "Get information on ranges receiving a high number of reads or writes." which we should probably mention here since it will allow folks to zero in on these hot ranges - this section should probably also add a link from that page back to here

@ajwerner since you're already on this review do you mind fact checking me on the above? also, do you have a concrete 1-3 recommendations to avoid read hot spots?